### ISSUE TITLE
Deploying chaincode fails when using OPENCHAIN_PEER_ADDRESSAUTODETECT

### ISSUE THREAD

**Issue by mastersingh24** [hyperledger-archives/fabric#417]
If you follow the devnet instructions and start multiple validating peer nodes (as docker containers) and then attempt to deploy chaincode, you will see

`vagrant@vagrant-ubuntu-trusty-64:/opt/gopath/src/github.com/openblockchain/obc-peer$ OPENCHAIN_PEER_ADDRESS=172.17.0.2:30303 ./obc-peer chaincode deploy -p github.com/openblockchain/obc-peer/openchain/example/chaincode/chaincode_example02 -v 0.0.1 -c '{"Function":"init", "Args": ["a","100", "b", "200"]}'
Error building chaincode: rpc error: code = 2 desc = "Error sending transactions to peer address=0.0.0.0:30303:  grpc: timed out trying to connect"`

**Comment by mastersingh24** [comment]
There are a couple of ways to fix this and I have implemented one on a separate branch, but it would be helpful if someone could explain the difference between **peer.listenAddress** and **peer.address** config params

**Comment by mastersingh24** [comment]
@muralisrini :  after our discussion, I did a little more research to figure out exactly what the issue is and what problem was trying to be solved by adding the listenAddress property.  From what I can tell:

**peer.addressAutoDetect** 

If this is set to **true**, then the code finds the first non loopback address to use

**peer.address** 

This property actually ends up being in in two ways.  First, if configured and **peer.addressAutoDetect** is **false**,it will be the address:port that the grpc endpoint uses.  The code also currently always passes this property to the chaincode containers as the address:port to use to communicate with the peer.  The default value in openchain.yaml is **0.0.0.0:30303**.  So if you set **peer.addressAutoDetect=true** or don't change it to an explicit IP address, then chaincode deploy will fail since chaincode will try to communicate with the peer using **0.0.0.0:30303** which of course is not valid as a client connection.  So even without the addition of the **peer.listenAddress** property, there was a bug when setting **peer.addressAutoDetect=true** as **peer.address** was still being passed to the chaincode containers.

**peer.listenAddress**

This was added in order to deal with the case where there is a "proxy" between chaincode containers and the obc-peer.  I use the term proxy loosely here as the specific reason was in the case where the obc-peer runs in a Docker container on one host and chaincode runs on a different host.  The current implementation use **peer.listenAddress** as the grpc endpoint and **peer.address** ends up being the "proxy" address which gets passed to the chaincode containers.

So basically I think we need to properly define the rule(s) for exactly which properties are required for different situations and which address / property should be sent to the chaincode containers

**Comment by kletkeman** [comment]
This is 2 weeks old and I just got it again today after syncing to the very latest code on master.

vagrant@vagrant-ubuntu-trusty-64:/opt/gopath/src/github.com/openblockchain/obc-peer$ OPENCHAIN_PEER_ADDRESS=172.17.0.2:30303 ./obc-peer chaincode deploy -p github.com/openblockchain/obc-peer/openchain/example/chaincode/chaincode_challenge_1 -c '{"Function":"init", "Args": ["COUNT1", "100"]}'
Error building chaincode: rpc error: code = 2 desc = "Error sending transactions to peer address=0.0.0.0:30303:  grpc: timed out trying to connect"

Is no one else running with multiple peers for testing? Or is there a work-around that has not yet been documented? Thx.

**Comment by corecode** [comment]
I believe I have a fix for this in my devel branch: corecode@f2166c2

**Comment by kletkeman** [comment]
@muralisrini, @ghaskins  Regarding the doc work mentioned in #373 for removal of the -v parameter ... 
@corecode Regarding the long delay in actually pushing a fix for the 0.0.0.0.30303 issue ... (by the way, thanks for responding and I will try to get things working with your fix) ...

For those of us trying to use / build upon obc for various reasons while we are also busy with other tasks, devnetsetup.md is a critical resource. Not only should the commands be tested and updated when the API changes (-v was removed for example), the end result should also be tested (for example, a successful 3 node network with deployment of example02.) 

Neither of these work right now on the latest master. And for the many who are or will be soon working with / on top of obc and not inside of obc, pulling dev branches is untidy and will not scale. Note that #451 was recently created, but the omissions have been around for a long time.

**Comment by srderson** [comment]
@kletkeman agreed. I think the best solution here is to integrate Travis CI with the GitHub repository and have it automatically setup a network and run tests (including documentation commands) against each PR. Ideally, the commands listed in the documentation and the commands run in the test originate from the same source. I think something like this is already being explored, but I will investigate and make sure it's on the list.

**Comment by kletkeman** [comment]
@srderson That sounds great. Sourcing from the same template for docs and automated tests should eliminate discrepancies between network and sandbox behaviors and also wasted time for both users and the dev team. Thanks!

**Comment by muralisrini** [comment]
@mastersingh24 : 
_" The code also currently always passes this property to the chaincode containers as the address:port to use to communicate with the peer. The default value in openchain.yaml is 0.0.0.0:30303. So if you set peer.addressAutoDetect=true or don't change it to an explicit IP address, then chaincode deploy will fail since chaincode will try to communicate with the peer using 0.0.0.0:30303 which of course is not valid as a client connection. So even without the addition of the peer.listenAddress property, there was a bug when setting peer.addressAutoDetect=true as peer.address was still being passed to the chaincode containers."_

Looking at code, that's not quite correct. chaincode code uses peer address from peer.GetPeerEndpoint func passed as arg to NewChaincodeSupport()...so chaincode was not unwittingly overriding address computed if peer.addressAutoDetect is set to true.  I think your eye may have caught the "peer.address" lines which are commented out ? 

For the basic issue with getting commands in DevnetSetup.md doc to work....

corecode@f2166c2 fix will work - basically refactor  peer.GetPeerEndpoint so we can use the computed peer address in a common manner.  That is definitely a first step so we can get DevnetSetup.md to work for everyone. I'll  work on fixing this and cleaning DevnetSetup.md doc. Will copy code from corecode@f2166c2. (I suspect @mastersingh24  fix likely does something close.)

Two issues beyond getting the basic doc instructions to work....
- To me the real issue is the way "peer.addressAutoDetect" works (and the reason for the hesitation with this issue). It grabs the address from the first interface that's not on the loopback interface and uses that.  That's quite arbitrary and we may need a better approach there. But let's handle that as a separate issue.
- Doc the need and rules regarding listen address (as @mastersingh24 mentions in comment)

**Comment by muralisrini** [comment]
Merged a fix using  code from corecode@f2166c2 along with changes to DevnetSetup.md doc.

**Comment by corecode** [comment]
Thanks!

**Comment by jeffgarratt** [comment]
@kletkeman is this still an issue for folks?
